### ISSUE TITLE
Add support in pack/unpack for missing file types

### ISSUE THREAD

**Issue by hickeng** [vmware/vic#5714]
Applies to #5700 

Currently only the basic file types are supported, but this could easily lead to infinite reads on `/dev/random` and definitely results in a failure to correctly export layers.
I suggest switching to `vendor\github.com\docker\docker\pkg\archive\archive.go` instead of the recently added `lib/archive` and modifying it as needed, or using as a reference for file type behaviours that are missing.
Changes likely needed if switching to it:
* addition of xattr headers to denote Added/Changed/Deleted - can this be done as a WhitelistFormat?
* translation of Import/Export/Rebase into `TarOptions`
* modify to take *os.File inputs instead of string paths - this allows for subsequent use of a non-mounted FileWalker style interface, e.g. nfs client.

Whichever approach we use, tether will need to have the same level of capability, or we will have to drop online `copyFrom`.

**Acceptance**
* encompass single file support & rename for import e.g. `docker cp file.txt cid:/foo.txt`
* copy entire container, including FIFOs, device nodes, etc
* must not duplicate data for hardlinks
* should include unit tests for filterspec generation and use by archive package

**Comment by morris-jason** [comment]
>encompass single file support & rename for import e.g. docker cp file.txt cid:/foo.txt
This is covered by valid implementations of ContainerStatPath.,

**Comment by jzt** [comment]
I've made lots of progress on this (thanks to some help and clarity from @emlin). The current state of the changes I've been making can be seen here:  
https://github.com/vmware/vic/compare/master...jzt:5714

Status:
Right now, tar file correctness is almost there with 3 known problems:

1.  When gzipped, the 4-byte (bytes 5-8) chunk of the gzip header that represents modification time does not match docker's, invalidating the blobsum calculation. @emlin and I suspect that the layer create time may be what is used here, but I haven't had a chance to test this as its a lower priority (this only really matters for `docker push` but I'm listing it here anyway)

2. An `sbin -> usr/sbin` link is created if `/sbin` doesn't already exist. I still have yet to find what is creating this link. The result is that in images like `busybox`, `sbin -> usr/sbin` shows up in the diff and it shouldn't be there. You can exec into the container, remove it, stop the container and diff, and its not there, but it comes back when you restart the container.

3. Whiteout files that are (were) hard links to other files are not accounted for yet. It's evident in the case of ubuntu:

Firstly, we can see that only a single image layer is incorrect between the two images: 

Filesizes (ours):
```
zach@arch:~/tmp/ours$ ls -lS
total 121256
-rw-r--r-- 1 zach zach 124115968 Aug 10 17:52 94160aca15bef16da0b738309df8008fcf878496cb1c07892f501c81a683c9e4.tar
-rw-r--r-- 1 zach zach     15872 Aug 10 17:52 9cb11a61de959ab1f5766e4f188e89c73e2b9016ee0acee7ad7f3ec0cd8f552d.tar
-rw-r--r-- 1 zach zach     14848 Aug 10 17:52 720755fc433ea66c1a4fe6bb7d74e0bf29576667e8693d3fcb45acc24460a4e1.tar
-rw-r--r-- 1 zach zach      5632 Aug 10 17:52 70229d75b264990c6d9941a5d9b486089dee364205009b84886843f8db98b9be.tar
-rw-r--r-- 1 zach zach      3072 Aug 10 17:52 09b1cf9f313801b24425d24aa928e4784e38a36fb59e52cf929addfefc538952.tar
-rw-r--r-- 1 zach zach      1024 Aug 10 17:52 58597429abbb866c8a6eac7eee6798ebf46025dc49b7c78c9d055d69ccbba77b.tar
```
Dockers:
```
zach@arch:~/tmp/theirs$ ls -lS
total 121252
-rw-r--r-- 1 zach zach 124115968 Aug 10 17:41 d5c6f90da05dc7e77d2e5fef63c341ab05ba2a03396ab5ae8f18814a7bbf5265.tar
-rw-r--r-- 1 zach zach     15872 Aug 10 17:41 1300883d87d5ae0582a4e0f3f86ad93985b18676e8d829f6f3b4990403fdf145.tar
-rw-r--r-- 1 zach zach     11776 Aug 10 17:41 c220aa3cfc1bff0c1df733eac869a66476ea35136c69d881da1e297ed8d0af2f.tar
-rw-r--r-- 1 zach zach      5632 Aug 10 17:41 2e9398f099dcc5a3e5521205e1f55ee5d7665e203193cf48b32553eabcca323c.tar
-rw-r--r-- 1 zach zach      3072 Aug 10 17:41 dc27a084064fee108f4d486469939bbfa6a5b16f6d6d3c6b5b2475e938f4c71e.tar
-rw-r--r-- 1 zach zach      1024 Aug 10 17:41 a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4.tar
```
SHA256 (ours):
```
zach@arch:~/tmp/ours$ sha256sum * |sort
25e0901a71b8c6a9df21590604a70517eb7b074071ef6af1033d50037baf3dd5  9cb11a61de959ab1f5766e4f188e89c73e2b9016ee0acee7ad7f3ec0cd8f552d.tar
49b2229a64e74e1535cf2b7c82a5cee3510e4dfe368921d17e15108e435d7086  720755fc433ea66c1a4fe6bb7d74e0bf29576667e8693d3fcb45acc24460a4e1.tar
5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef  58597429abbb866c8a6eac7eee6798ebf46025dc49b7c78c9d055d69ccbba77b.tar
8aa4fcad5eeb286fe9696898d988dc85503c6392d1a2bd9023911fb0d6d27081  94160aca15bef16da0b738309df8008fcf878496cb1c07892f501c81a683c9e4.tar
9c42c2077cdea659ac116f148b63ded203d0e83f017accb6d7987377d1363673  70229d75b264990c6d9941a5d9b486089dee364205009b84886843f8db98b9be.tar
a09947e71dc0d591901870f2fd3c18565339b960f6bb32793d14604a85d67393  09b1cf9f313801b24425d24aa928e4784e38a36fb59e52cf929addfefc538952.tar
```
Theirs:
```
zach@arch:~/tmp/theirs$ sha256sum *|sort
25e0901a71b8c6a9df21590604a70517eb7b074071ef6af1033d50037baf3dd5  1300883d87d5ae0582a4e0f3f86ad93985b18676e8d829f6f3b4990403fdf145.tar
5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef  a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4.tar
625c7a2a783b4736cf488efd1fafc41736b9998ec087e20da946c30522ec9ad7  c220aa3cfc1bff0c1df733eac869a66476ea35136c69d881da1e297ed8d0af2f.tar
8aa4fcad5eeb286fe9696898d988dc85503c6392d1a2bd9023911fb0d6d27081  d5c6f90da05dc7e77d2e5fef63c341ab05ba2a03396ab5ae8f18814a7bbf5265.tar
9c42c2077cdea659ac116f148b63ded203d0e83f017accb6d7987377d1363673  2e9398f099dcc5a3e5521205e1f55ee5d7665e203193cf48b32553eabcca323c.tar
a09947e71dc0d591901870f2fd3c18565339b960f6bb32793d14604a85d67393  dc27a084064fee108f4d486469939bbfa6a5b16f6d6d3c6b5b2475e938f4c71e.tar
```

So we can see that our layer `720755fc` doesn't fully match with docker's `c220aa3c`:
```
zach@arch:~/tmp/ours$ tar tvf 720755fc433ea66c1a4fe6bb7d74e0bf29576667e8693d3fcb45acc24460a4e1.tar >a
zach@arch:~/tmp/ours$ tar tvf ../theirs/c220aa3cfc1bff0c1df733eac869a66476ea35136c69d881da1e297ed8d0af2f.tar >b
zach@arch:~/tmp/ours$ diff a b
5,21c5,21
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_main_binary-amd64_Packages
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_main_i18n_Translation-en
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_restricted_binary-amd64_Packages
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_restricted_i18n_Translation-en
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_InRelease
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_main_binary-amd64_Packages
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_main_i18n_Translation-en
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_restricted_binary-amd64_Packages
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_restricted_i18n_Translation-en
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.lock
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.partial
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_InRelease
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_main_binary-amd64_Packages
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_main_i18n_Translation-en
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_restricted_binary-amd64_Packages
< ---------- 0/0               0 2017-08-10 17:52 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_restricted_i18n_Translation-en
---
> -r--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_main_binary-amd64_Packages link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_main_i18n_Translation-en link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_restricted_binary-amd64_Packages link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_restricted_i18n_Translation-en link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_InRelease link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_main_binary-amd64_Packages link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_main_i18n_Translation-en link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_restricted_binary-amd64_Packages link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial_restricted_i18n_Translation-en link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.lock link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.partial link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_InRelease link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_main_binary-amd64_Packages link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_main_i18n_Translation-en link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_restricted_binary-amd64_Packages link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
> hr--r--r-- 0/0               0 2017-08-10 15:13 var/lib/apt/lists/.wh.security.ubuntu.com_ubuntu_dists_xenial-security_restricted_i18n_Translation-en link to var/lib/apt/lists/.wh.archive.ubuntu.com_ubuntu_dists_xenial-updates_InRelease
```

In that sense, all that remains to be done is something like:
- during header creation, we stat the original (now deleted) file on the old path to find out if its a hard link
- add it to our list of seen links as we do with other hard links:  https://github.com/vmware/vic/compare/master...jzt:5714#diff-39ee8952697fe56877c42008878327c7R283

There may be some filename/path formatting involved since they are whiteout files linking to other whiteout files, but I think that can be covered generally by a very similar implementation to what is there now. There also needs some attention paid to the timestamp in those cases (the case of whiteout files in general) so that they match docker's. I suspect the original file's timestamp can be used in the whiteout header.

**Comment by jzt** [comment]
So after some extensive research with @emlin, we've discovered that the whiteout hard links in these layers are not actually hard links at all. 

When docker does the diff between the layer on which these files originally existed and the child layer reflecting their removal, they consider these files to be hard links when creating the layer archive. We have mounted the disk containing these files and inspected them by hand and verified that they are indeed not hard links. We suspect this may be a bug in docker code. If it isn't a bug, it could be something aufs related, or perhaps some type of optimization.

In the end, though, even though the tar archives technically don't match from a `docker push` perspective, the end result  will be the same - the child layer will not contain the deleted files. So long as we are not concerned about docker push, it should not be a blocker to 1.2. This issue has been captured here:  https://github.com/vmware/vic/issues/5997

Similarly, upon further considering, #2 [above](https://github.com/vmware/vic/issues/5714#issuecomment-321699707) is not related to this issue. It is caused by something unrelated to diff (which is merely reflecting what exists on the filesystem and changes between layers). I new issue capturing this bug is here:  https://github.com/vmware/vic/issues/5996

That being said, all acceptance criteria in this issue has been met.
